### PR TITLE
Add Trim operation

### PR DESCRIPTION
**Pull request by mateusztomaszewski**
## What:

* Add Trim operation 

## Why:

*  To extended basic operations set

## Example:

* See unit tests

### CODE REVIEW

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package string
+
+import operation.StandardLogicOperation
+import utils.asList
+import utils.secondOrNull
+import utils.thirdOrNull
+
+object Trim : StandardLogicOperation, StringUnwrapStrategy {
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:8]: Add it to the stlib operations provider

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    override fun evaluateLogic(expression: Any?, data: Any?): Any?  {
+        return when (val firstItem = expression.asList.firstOrNull()) {
+            is String -> firstItem.trimElements(' ', TrimMode.BothEnds)
+            is List<*> -> {
+                val text = firstItem.firstOrNull()
+                val sign = firstItem.secondOrNull() ?: " "
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:14]:
```suggestion
                val sign = firstItem.secondOrNull().orEmpty()
```

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            }
+            else -> null
+        }
+
+    private fun modeBasedTrim(mode: TrimMode,
+                              start: (() -> Any?),
+                              end: (() -> Any?),
+                              bothEnds: (() -> Any?)
+    ) =
+        when (mode) {
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:50]: strange line break

File: operations-stdlib/src/commonTest/kotlin/string/TrimTest.kt [hunk 4]
```diff
@@ -0,0 +1,133 @@
+import JsonLogicResult.Failure
+import JsonLogicResult.Success
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.datatest.withData
+import io.kotest.matchers.shouldBe
+import string.Trim
+
+class TrimTest : FunSpec({
```
Comment (KTGypsy) [inline operations-stdlib/src/commonTest/kotlin/string/TrimTest.kt:8]: adding some more test cases where you use **data** and/or complex expressions with some other operations(var for example) would be great

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+import operation.StandardLogicOperation
+import utils.asList
+import utils.secondOrNull
+import utils.thirdOrNull
+
+object Trim : StandardLogicOperation, StringUnwrapStrategy {
+    override fun evaluateLogic(expression: Any?, data: Any?): Any?  {
+        return when (val firstItem = expression.asList.firstOrNull()) {
+            is String -> firstItem.trimElements(' ', TrimMode.BothEnds)
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:11]: We discussed it on slack that we dont want to use defaults in the library. When some arguments are missing we should treat it as a wrong input and usually return null. What we decided is to declare operation defaults on backend/dsl library.

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    private fun String?.toTrimMode() = when (this) {
+        "start" -> TrimMode.Start
+        "end" -> TrimMode.End
+        "bothEnds" -> TrimMode.BothEnds
+        else -> null
+    }
+
+    private fun Any?.trimElements(char: Char, mode: TrimMode) =
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:32]: this fuction might be shortened to
image

**Review by KTGypsy — CHANGES_REQUESTED** [review]

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        when (this) {
+            is String -> {
+                modeBasedTrim(
+                    mode = mode,
+                    start = { trimStart(char) },
+                    end = { trimEnd(char) },
+                    bothEnds = { trim(char) }
+                )
+            }
+            else -> null
```
Comment (mkrogulskiallegro2) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:42]: This result is not being included in tests -
 case of having list as first argument ->
`operatorName to listOf(listOf(listOf(" cat "), " ", "bothEnds"))` - of course it should fail

same goes for making call from Data, which is not tested either (it should work out of the box, but having single verifier in your code won't hurt). 

`            TestInput(
                expression = mapOf(
                    operatorName to listOf(listOf(listOf(" cat "), " ", "bothEnds"))
                ),
                result = Success("cat")
            ),
            TestInput(
                expression = mapOf(
                    operatorName to listOf(listOf(mapOf("var" to "key"), " ", "bothEnds"))
                ),
                data = mapOf("key" to listOf(" cat ", " cat ")),
                result = Success("cat")
            ),`

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        }
+
+    private fun modeBasedTrim(mode: TrimMode,
```
Comment (mkrogulskiallegro2) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:45]: actually.. why make this so complicated? With highe order functions, providing from other place and not just

```
    private fun Any?.trimElements(char: Char, mode: TrimMode) =
        when (this) {
            is String -> mode.trimByMode(this, char)
            else -> null
        }
```

```
    private fun TrimMode.trimByMode(target: String, char: Char) =
        when (this) {
            TrimMode.Start -> target.trimStart(char)
            TrimMode.End -> target.trimEnd(char)
            TrimMode.BothEnds -> target.trim(char)
        }
```

less 'any' usage ;)

**Review by mkrogulskiallegro2 — CHANGES_REQUESTED** [review]

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt (hunk 1 above)
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt:9]: 👍 done

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt (hunk 5 above)
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt:12]: 👍

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt (hunk 7 above)
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt]: 👍 I have fixed it in the new commit

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt (hunk 8 above)
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt]: 👍 I changed it as @KTGypsy suggested

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    private fun String?.toTrimMode() = when (this) {
+        "start" -> TrimMode.Start
+        "end" -> TrimMode.End
+        "bothEnds" -> TrimMode.BothEnds
+        else -> null
+    }
+
+    private fun Any?.trimElements(char: Char, mode: TrimMode) =
+        when (this) {
```
[the remark this answers is not in the document]
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt]: Done

File: operations-stdlib/src/commonTest/kotlin/string/TrimTest.kt (hunk 4 above)
Comment (KTGypsy) [reply operations-stdlib/src/commonTest/kotlin/string/TrimTest.kt:8]: I think its fine now, tests for operators are almost infinite. The more the better :D

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            TrimMode.Start -> this.text.trimStart(char)
+            TrimMode.End -> this.text.trimEnd(char)
+            TrimMode.BothEnds -> this.text.trim(char)
+        }
+
+    private fun String?.toTrimMode() = when (this) {
+        "start" -> TrimMode.Start
+        "end" -> TrimMode.End
+        "bothEnds" -> TrimMode.BothEnds
+        else -> throw IllegalStateException("Invalid TrimMode value")
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:40]: just return null, engine will convert the result and return proper JsonLogicResult object

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+object Trim : StandardLogicOperation, StringUnwrapStrategy {
+    private const val ARGUMENTS_ARG_INDEX = 0
+    private const val TEXT_ARG_INDEX = 0
+    private const val CHAR_ARG_INDEX = 1
+    private const val MODE_ARG_INDEX = 2
+
+    override fun evaluateLogic(expression: Any?, data: Any?): Any? =
+        expression.asList.toOperationArguments()?.invokeTrim()
+
+    private fun TrimArguments.invokeTrim() = (this as? TrimArguments)?.modeBasedTrim(mode, char)
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:15]: To improve readability move this function under all of the rest

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+    private fun List<Any?>.toOperationArguments(): TrimArguments? = runCatching {
+        val arguments = get(ARGUMENTS_ARG_INDEX).asList
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:18]: the input in your unit tests is nested too deeply and thats why you need second unwrapping. I think its unnecessary and unit tests should look like:
`operatorName to listOf("_cat_", "_", "start")`

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 13]
```diff
@@ -0,0 +1,54 @@
+package string
+
+import operation.StandardLogicOperation
+import utils.asList
+
+object Trim : StandardLogicOperation, StringUnwrapStrategy {
+    private const val ARGUMENTS_ARG_INDEX = 0
```
Comment (KTGypsy) [inline operations-stdlib/src/commonMain/kotlin/string/Trim.kt:7]: if you apply changes from comments below this will become unused

Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt]: Done

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt (hunk 12 above)
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt]: Done

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt (hunk 11 above)
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt]: Done

**Review by KTGypsy — DISMISSED** [review]

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    )
+
+    private fun String?.toTrimMode() = when (this) {
+        "start" -> TrimMode.Start
+        "end" -> TrimMode.End
+        "bothEnds" -> TrimMode.BothEnds
+        else -> throw IllegalStateException("Invalid TrimMode value")
+    }
+
+    private fun TrimArguments.invokeTrim() = (this as? TrimArguments)?.modeBasedTrim(mode, char)
```
[the remark this answers is not in the document]
Comment (PaulinaSadowskaAllegro) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt:32]: Or maybe even remove one of the methods? 
```
    private fun TrimArguments.invokeTrim() = when (mode) {
        TrimMode.Start -> this.text.trimStart(char)
        TrimMode.End -> this.text.trimEnd(char)
        TrimMode.BothEnds -> this.text.trim(char)
    }
```

**Review by PaulinaSadowskaAllegro — DISMISSED** [review]

File: operations-stdlib/src/commonMain/kotlin/string/Trim.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (mateusztomaszewski) [reply operations-stdlib/src/commonMain/kotlin/string/Trim.kt:32]: 👌 You're right, I'm correcting.

**Review by PaulinaSadowskaAllegro — APPROVED** [review]

**Review by KTGypsy — APPROVED** [review]

**Pull request merged**